### PR TITLE
Fault Tolerant Manual: Add loading to reload the states

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Find the PoC here: #10568

Parts of #10431

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @justusschock @awaelchli @ninginthecloud

### CODE REVIEW

File: pytorch_lightning/utilities/auto_restart.py [hunk 1]
```diff
@@ -564,35 +564,66 @@ def _add_capture_metadata_collate(dataloader: DataLoader) -> None:
...
+
+        if not hasattr(dataset, "load_state_dict"):
+            return
+
+        dataset_state = {
+            worker_id: state_dict["state"][worker_id]["dataset_state"][worker_id]
+            for worker_id in state_dict["state"].keys()
+        }
+
+        dataset.load_state_dict(_rotate_worker_indices(dataset_state, latest_worker_id, num_workers))
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:626]: it could maybe be confusing that we rotate the worker_indicies here but in automatic mode we do it inside the dataset (when calling load_state_dict) .

File: pytorch_lightning/utilities/auto_restart.py [hunk 2]
```diff
@@ -564,35 +564,66 @@ def _add_capture_metadata_collate(dataloader: DataLoader) -> None:
...
+
+    elif fault_tolerant_mode.is_manual:
+
+        latest_worker_id = state_dict["latest_worker_id"]
+        num_workers = state_dict["state"][latest_worker_id]["num_workers"]
+        sampler_state = state_dict["state"][latest_worker_id]["sampler_state"]
+        if sampler_state:
+            for k in sampler_state:
+                obj = getattr(dataloader, k)
+                if not isinstance(obj, _SupportsStateDict):
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:611]: maybe check for callable(obj) too?

File: pytorch_lightning/utilities/auto_restart.py [hunk 3]
```diff
@@ -564,35 +564,66 @@ def _add_capture_metadata_collate(dataloader: DataLoader) -> None:
...
+                    raise MisconfigurationException(
+                        f"The DataLoader attribute should have a `load_state_dict` method. Found {obj}"
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:613]: calling it attribute may be misleading. Usually the dataloader is not an attribute

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/auto_restart.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:612]: Good catch !

File: pytorch_lightning/utilities/auto_restart.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:625]: Added a comment.

File: pytorch_lightning/utilities/auto_restart.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:610]: The object doesn't need to be callable. It just need to matching the interface  `_SupportsStateDict `. Here, we raise an exception as it was collected previously and the object matched the `_SupportsStateDict` interface.

```py
    def _store_sampler_state(self) -> None:
        """This function is used to extract the sampler states if any."""
        sampler_state = {
            k: v.state_dict()
            for k, v in self._loader.__dict__.items()
            if isinstance(v, _SupportsStateDict) and k != "dataset"
        }
        self.__accumulate_state(sampler_state)
```

File: pytorch_lightning/utilities/auto_restart.py [hunk 4]
```diff
@@ -564,35 +563,69 @@ def _add_capture_metadata_collate(dataloader: DataLoader) -> None:
...
+            raise MisconfigurationException("This shouldn't happen. Please, open an issue on PyTorch Lightning Github.")
+
+    elif fault_tolerant_mode.is_manual:
+
+        # In manual mode, we don't wrap the user objects with `CaptureMapDataset` or `CaptureIterableDataset`
+        # therefore, we need to reload the states manually.
+
+        latest_worker_id = state_dict["latest_worker_id"]
+        num_workers = state_dict["state"][latest_worker_id]["num_workers"]
+        sampler_state = state_dict["state"][latest_worker_id]["sampler_state"]
```
Comment (ethanwharris) [inline pytorch_lightning/utilities/auto_restart.py:609]: Should this maybe be:
```suggestion
        sampler_state = state_dict["state"][latest_worker_id].get("sampler_state", None)
```

File: pytorch_lightning/utilities/auto_restart.py [hunk 5]
```diff
@@ -564,35 +563,69 @@ def _add_capture_metadata_collate(dataloader: DataLoader) -> None:
...
+        if sampler_state:
+            for k in sampler_state:
```
Comment (ethanwharris) [inline pytorch_lightning/utilities/auto_restart.py:611]: What are the possible values here? If `sampler_state` can be None then maybe `if sampler_state is not None` would be cleaner. Otherwise if it can only be an empty dict then the if would not be needed.

**Review by ethanwharris — APPROVED** [review]
LGTM, I have some minor confusion about the logic

File: pytorch_lightning/utilities/auto_restart.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:609]: Done !

File: pytorch_lightning/utilities/auto_restart.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:611]: Renamed `dataloader_attr_name` instead of k. This would be all attribute name of all the attributes within the DataLoader matching the `_SupportsStateDict` API.

Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:611]: Added a comment.

**Review by kaushikb11 — APPROVED** [review]
LGTM!

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/utilities/auto_restart.py [hunk 6]
```diff
@@ -564,35 +563,71 @@ def _add_capture_metadata_collate(dataloader: DataLoader) -> None:
     )
 
 
-def reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, Any]) -> None:
+def _reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, Any]) -> None:
```
Comment (carmocca) [inline pytorch_lightning/utilities/auto_restart.py:566]: For the sake of good code practices, this should be

```python
# loop.py
ft_mode = self.trainer.state._fault_tolerant_mode
if ft_mode.is_automatic:
    _reload_automatic_dataloader_state()
elif ft_mode.is_manual:
    _reload_manual_dataloader_state()
```

Instead of checking the mode inside and have two completely different code paths for each mode. This aids in testing and separates responsabilities.

Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py]: Done !

File: pytorch_lightning/utilities/auto_restart.py [hunk 7]
```diff
@@ -564,38 +563,90 @@ def _add_capture_metadata_collate(dataloader: DataLoader) -> None:
...
+    num_workers = state_dict["state"][latest_worker_id]["num_workers"]
+    sampler_state = state_dict["state"][latest_worker_id].get("sampler_state", None)
+    if sampler_state:
+        # `sampler_state` keys contain all the DataLoader attribute names
+        # which matched `_SupportsStateDict` API interface while collecting the `state_dict`.
+        for dataloader_attr_name in sampler_state:
+            obj = getattr(dataloader, dataloader_attr_name)
+            if not isinstance(obj, _SupportsStateDict):
+                raise MisconfigurationException(
+                    f"The DataLoader attribute {dataloader_attr_name}:{obj} should have a `load_state_dict` method."
```
Comment (carmocca) [inline pytorch_lightning/utilities/auto_restart.py:618]:
```suggestion
                    f"The DataLoader attribute {dataloader_attr_name}: {obj} should have a `load_state_dict` method."
```

File: pytorch_lightning/utilities/auto_restart.py [hunk 8]
```diff
@@ -564,38 +563,90 @@ def _add_capture_metadata_collate(dataloader: DataLoader) -> None:
...
+
+    latest_worker_id = state_dict["latest_worker_id"]
+    num_workers = state_dict["state"][latest_worker_id]["num_workers"]
+    sampler_state = state_dict["state"][latest_worker_id].get("sampler_state", None)
+    if sampler_state:
+        # `sampler_state` keys contain all the DataLoader attribute names
+        # which matched `_SupportsStateDict` API interface while collecting the `state_dict`.
+        for dataloader_attr_name in sampler_state:
+            obj = getattr(dataloader, dataloader_attr_name)
+            if not isinstance(obj, _SupportsStateDict):
```
Comment (carmocca) [inline pytorch_lightning/utilities/auto_restart.py:616]: The comment above says:

"`sampler_state` keys contain all the DataLoader attribute names which matched `_SupportsStateDict` API interface while collecting the `state_dict`."

doesn't that mean that this `isinstance` check is unnecessary? The sampler state keys are only there if it supports state dict.

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/utilities/auto_restart.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:618]: Done !

File: pytorch_lightning/utilities/auto_restart.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:616]: Yes, but the function should be independent from the collection. We need to ensure the same dataloader is being provided when reloading.
